Tests/LinearSolvers/ABecLaplacian_C/MyTest.cpp

```diff
@@ -460,6 +466,86 @@ MyTest::solveNodeABecLaplacian ()
     }
 }
 
+void
+MyTest::solveABecLaplacianGMRES ()
+{
+    LPInfo info;
+    info.setMaxCoarseningLevel(0);
+
+    const auto tol_rel = Real(1.e-6);
+    const auto tol_abs = Real(0.0);
+
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(composite_solve == false,
+       "solveABecLaplacianGMRES does not support composite solve");
+
+    const auto nlevels = static_cast<int>(geom.size());
+
+    for (int ilev = 0; ilev < nlevels; ++ilev)
+    {
+        MLABecLaplacian mlabec({geom[ilev]}, {grids[ilev]}, {dmap[ilev]}, info);
+
+        mlabec.setMaxOrder(linop_maxorder);
+
+        // This is a 3d problem with homogeneous Neumann BC
+        mlabec.setDomainBC({AMREX_D_DECL(LinOpBCType::Neumann,
+                                         LinOpBCType::Neumann,
+                                         LinOpBCType::Neumann)},
+            {AMREX_D_DECL(LinOpBCType::Neumann,
+                          LinOpBCType::Neumann,
+                          LinOpBCType::Neumann)});
+
+        if (ilev > 0) {
+            mlabec.setCoarseFineBC(&solution[ilev-1], ref_ratio);
+        }
+
+        // for problem with pure homogeneous Neumann BC, we could pass a nullptr
+        mlabec.setLevelBC(0, nullptr);
+
+        mlabec.setScalars(ascalar, bscalar);
+
+        mlabec.setACoeffs(0, acoef[ilev]);
+
+        Array<MultiFab,AMREX_SPACEDIM> face_bcoef;
+        for (int idim = 0; idim < AMREX_SPACEDIM; ++idim)
+        {
+            const BoxArray& ba = amrex::convert(bcoef[ilev].boxArray(),
+                                                IntVect::TheDimensionVector(idim));
+            face_bcoef[idim].define(ba, bcoef[ilev].DistributionMap(), 1, 0);
+        }
+        amrex::average_cellcenter_to_face(GetArrOfPtrs(face_bcoef),
+                                          bcoef[ilev], geom[ilev]);
+        mlabec.setBCoeffs(0, amrex::GetArrOfConstPtrs(face_bcoef));
+
+        MultiFab res(rhs[ilev].boxArray(), rhs[ilev].DistributionMap(), 1, 0);
+
+        MLMG mlmg(mlabec);
+        mlmg.setVerbose(verbose);
+        mlmg.apply({&res}, {&solution[ilev]}); // res = L(sol)
+
+        MultiFab::Subtract(res, rhs[ilev], 0, 0, 1, 0); // now res = L(sol) - rhs
+
+        MultiFab cor(rhs[ilev].boxArray(), rhs[ilev].DistributionMap(), 1, 0);
+
+        using M = GMRESMLMG;
+        M mat(mlmg);
+        mat.usePrecond(true);
+
+        GMRES<MultiFab,M> gmres;
+        gmres.setVerbose(verbose);
+        gmres.define(mat);
+        gmres.solve(cor, res, tol_rel, tol_abs); // solve L(cor) = res
+
+        MultiFab::Subtract(solution[ilev], cor, 0, 0, 1, 0);
+
+        mlmg.apply({&res}, {&solution[ilev]}); // res = L(sol)
+        MultiFab::Subtract(res, rhs[ilev], 0, 0, 1, 0); // now res = L(sol) - rhs
+        if (verbose) {
+            amrex::Print() << "Final residual = " << res.norminf(0)
+                           << " " << res.norm1(0) << " " << res.norm2(0) << std::endl;
+        }
+    }
+}
+
```

Comment:
Is this loop eventually intended to be added to either `GRESMLMG` or `GMRES`?

Could we perhaps generalize this to replace `gmres` with other solver types that have thus far been used only as bottom solvers? Some examples would include the `CG` and `BiCGStab` methods.

As is, `MLMG` seems to force the user to always use the smoother for the non-bottom steps, which uses up the overwhelming computational time. It would be nice if one could make that an option. I imagine the boundary conditions would have to handled more carefully in the case of the non-smoother methods.

It would also be nice if one could select `GMRES` as an option in `MLMG` as the bottom solver.

Reply:
Thanks for your comments. This PR is just first the step of adding GMRES support. Yes, eventually we would like to be able to use GMRES as a bottom solver in MLMG and use MLMG as a multigrid preconditioner for GMRES. The GMRES class will probably be more or less the same, but we need to make some changes to MLMG and the interface class GMRESMLMG.

> As is, MLMG seems to force the user to always use the smoother for the non-bottom steps,

MLMG calls `linop.smooth`. So it's up to the linear operator class on what to do. Nevertheless, I think the name `smooth` is appropriate, because that's how multigrid works. The relaxation step is supposed to smooth out the high-frequency modes of the error. I don't know what is in your mind as not a "smoother". I suspect whatever that is, it needs to behave like a smoother in terms of eliminating high-frequency errors for it to be effective.